userspace/libsinsp/test/public_sinsp_API/sinsp_logger.cpp

```diff
@@ -67,3 +263,262 @@ TEST(sinsp_logger, get_set_severity)
+	ASSERT_EQ(err, "");
+	ASSERT_EQ(file, "");
+}
+
+/**
+ * With a standard ouput logging sink enabled (and the sinsp_logger::OT_NOTS
+ * enabled), emitted logs should be written only to standard output, and those
+ * logs do not contain the timestamp.
+ */
+TEST_F(sinsp_logger_test, log_standard_output_nots)
+{
+	std::string out;
+	std::string err;
+	std::string file;
+
+	libsinsp_logger()->reset();
+	libsinsp_logger()->add_stdout_log();
+	libsinsp_logger()->disable_timestamps();
+
+	ASSERT_EQ(libsinsp_logger()->get_log_output_type(),
+	          (sinsp_logger::OT_STDOUT | sinsp_logger::OT_NOTS));
+
+	generate_log(DEFAULT_MESSAGE, out, err, file, sinsp_logger::SEV_FATAL);
+
+	// The logging API appends a newline
+	const std::string expected = DEFAULT_MESSAGE + "\n";
+
+	ASSERT_EQ(expected, out);
+	ASSERT_EQ(err, "");
+	ASSERT_EQ(file, "");
+}
+
+/**
+ * With stderr logging sink enabled, emitted logs should be written only to
+ * standard error.
+ */
+TEST_F(sinsp_logger_test, log_standard_error)
+{
+	std::string out;
+	std::string err;
+	std::string file;
+
+	libsinsp_logger()->reset();
+	libsinsp_logger()->add_stderr_log();
+	ASSERT_EQ(libsinsp_logger()->get_log_output_type(), sinsp_logger::OT_STDERR);
+
+	generate_log(DEFAULT_MESSAGE, out, err, file, sinsp_logger::SEV_FATAL);
+
+	ASSERT_EQ(out, "");
+	ASSERT_NE(err.find(DEFAULT_MESSAGE), std::string::npos);
+	ASSERT_EQ(file, "");
+}
+
+/**
+ * With file logging sink enabled, emitted logs should be written only to the
+ * file.
+ */
+TEST_F(sinsp_logger_test, log_file)
+{
+	const std::string filename = "/tmp/ut.out";  // FIXME
+	std::string out;
+	std::string err;
+	std::string file;
+
+	libsinsp_logger()->reset();
+	libsinsp_logger()->add_file_log(filename);
+	ASSERT_EQ(libsinsp_logger()->get_log_output_type(), sinsp_logger::OT_FILE);
+
+	generate_log(DEFAULT_MESSAGE, out, err, file, sinsp_logger::SEV_FATAL, filename);
+
+	ASSERT_EQ(out, "");
+	ASSERT_EQ(err, "");
+	ASSERT_NE(file.find(DEFAULT_MESSAGE), std::string::npos);
+}
+
+/**
+ * With a callback logging sink enabled, emitted logs should be written only to
+ * the callback.
+ */
+TEST_F(sinsp_logger_test, log_callback)
+{
+	libsinsp_logger()->reset();
+	libsinsp_logger()->add_callback_log(log_callback_fn);
+	ASSERT_EQ(libsinsp_logger()->get_log_output_type(), sinsp_logger::OT_CALLBACK);
+
+	libsinsp_logger()->log(DEFAULT_MESSAGE, sinsp_logger::SEV_FATAL);
+
+	ASSERT_NE(get_callback_output().find(DEFAULT_MESSAGE), std::string::npos);
+}
+
+TEST_F(sinsp_logger_test, log_stderr_multithreaded)
+{
+	const size_t NUM_THREADS = 5;
+	const std::string message = "123456789";  // 9 characters
+	const size_t NUM_LOGS = 80;
+	const size_t NUM_SUBSTRINGS = NUM_THREADS * NUM_LOGS;
+
+	// 5 threads *
+	// 80 logs of each 9 characters (plus a '\n') = 10 characters *
+	// = 4000 characters, which should be less than BUFFER_SIZE
+	ASSERT_TRUE((NUM_SUBSTRINGS * (message.size() + 1)) < (BUFFER_SIZE - 1));
+
+	libsinsp_logger()->reset();
```

Comment:
LTGM!

Would it make more sense to invoke `libsinsp_logger()->reset();` inside `sinsp_logger_test::SetUp()`, since we do that at the start of each test?

Reply:
AFAIK the SetUp is run once at the start and not once for each test but I could be wrong. Let me check.

Reply:
You were right :)